### ISSUE TITLE
Uncaught (in promise) TypeError: Cannot add property areEquivalent_Eq_91_2_34

### ISSUE THREAD

**Issue by colvint** [ag-grid/ag-grid-react#81]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid-react/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
My grid is throwing an `Uncaught (in promise) TypeError: Cannot add property areEquivalent_Eq_91_2_34, object is not extensible` error when rows within rowData are updated by Redux. The exception is being thrown [here](https://github.com/ceolter/ag-grid-react/blob/8d5a4f45a907dc121c01ec84fe6eca12039322dd/lib/agGridReact.js#L121)

In my case, `a` and `b` in the code are all the rows in my grid:
![image](https://user-images.githubusercontent.com/144203/28548062-ea84e610-709f-11e7-9997-feeaadbb831e.png)

**Expected behavior**
I'm simply updating one or more rows in a list on the server. The server returns the updated row data into the redux store. I would expect (I think this just used to work) the changes to flow from the redux store into the props of my grid. The affected rows in the grid should reflect those changes. Note this isn't an insert -- just an update of existing row data.

**Minimal reproduction of the problem with instructions**

```javascript
// my grid config
<AgGridReact
            columnDefs={this.state.columnDefs}
            gridOptions={{
              rowSelection: "multiple",
              enableColResize: true,
              suppressRowClickSelection: true,
              onGridReady: this.handleGridReady,
              onSelectionChanged: this.handleSelectionChanged
            }}
            rowData={enrolledStudents}
            rowHeight={LIST_ROW_HEIGHT}
            enableColResize="true"
            enableSorting="true"
            enableFilter="true"
          />
        </div>
```

**Please tell us about your environment:**
OSX 10.11.6
```
    "ag-grid": "^12.0.0",
    "ag-grid-enterprise": "^12.0.0",
    "ag-grid-react": "^12.0.0",
    "apollo-client": "^1.9.0-1",
...
    "react": "^15.6.1",
    "react-apollo": "^1.4.7",
    "react-dom": "^15.6.1",
    "react-redux": "^5.0.5",
    "react-router": "3.0.5",
    "redux": "^3.7.2",
```

* **ag-grid-react version:** 2.0.X
12.0.0

* **ag-Grid/ag-Grid-Enterprise version:** 2.0.X
12.0.0

* **React version:** 2.0.X
15.6.1

* **Browser:**
Latest Chrome: Version 59.0.3071.115 (Official Build) (64-bit)

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
ES6

**Comment by colvint** [comment]
Hmm, I'm getting nearly the same error without even updating rows. Something else must be afoot. I'll report back once I've figured it out.

**Comment by seanlandsman** [comment]
Hi

I can see that this is thrown when in strict mode or when an object is not extensible - I cannot reproduce this in the test cases I have at the moment.

If you can provide something that reproduces this I'd be happy to take a look and make the necessary changes

thanks

**Comment by colvint** [comment]
@seanlandsman I'm not sure how to create a plunkr which completely models my setup. I've got redux feeding in my `rowData`, etc. I did just confirm the issue goes away if i downgrade to v`10.0.0`.

**Comment by seanlandsman** [comment]
Hi - I dont think I need a plunker as such, a github repo or similar would help

I'll see if I can reproduce it in our redux example here https://github.com/ceolter/ag-grid-react-example/tree/master/src/simpleReduxExample

Maybe you can take a look at that example and see what differences in approach I have and you do?

thanks

**Comment by colvint** [comment]
The repo i'm working on is private so i can't share it unfortunately. My approach is very similar to the simple redux example, *except* that I'm also using react router. I mention that because the exception is thrown when I change routes and, presumably the ag-grid gets un/mounted. You might be able to repro if you have a redux-fed ag-grid on the current route, and then navigate to another route with an ag-grid.

**Comment by theofanis** [comment]
Same problem here, but I can't share code as well :(

One property in my grid data model is `JSX.Element` type (using `typescript`), which is a simple `<span>` element rendering a `string` value.

Maybe this helps?
![image](https://user-images.githubusercontent.com/6344512/28779036-5f9b7620-760a-11e7-842f-0cd6e9d4a762.png)

**Comment by colvint** [comment]
@theofanis Downgrading to `10.0.0` is my temporary workaround.

**Comment by seanlandsman** [comment]
@colvint I'll try look at this next week - just fyi

**Comment by seanlandsman** [comment]
@ceolter @theofanis 

I think I have a fix here, but as I can't reproduce it I'd like you guys to confirm it works

Can you please replace the contents of node_modules/ag-grid-react/lib/agGridReact.js with the following:


`// ag-grid-react v12.0.0
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var reactFrameworkFactory_1 = require("./reactFrameworkFactory");
var reactFrameworkComponentWrapper_1 = require("./reactFrameworkComponentWrapper");
var DOM = require("react-dom-factories");
var react_1 = require("react");
var PropTypes = require("prop-types");
var AgGrid = require("ag-grid");
var AgGridReact = (function (_super) {
    __extends(AgGridReact, _super);
    function AgGridReact(props, state) {
        var _this = _super.call(this, props, state) || this;
        _this.props = props;
        _this.state = state;
        return _this;
    }
    AgGridReact.prototype.render = function () {
        var _this = this;
        return DOM.div({
            style: this.createStyleForDiv(),
            ref: function (e) {
                _this.eGridDiv = e;
            }
        });
    };
    AgGridReact.prototype.createStyleForDiv = function () {
        var style = { height: '100%' };
        // allow user to override styles
        var containerStyle = this.props.containerStyle;
        if (containerStyle) {
            Object.keys(containerStyle).forEach(function (key) {
                style[key] = containerStyle[key];
            });
        }
        return style;
    };
    AgGridReact.prototype.componentDidMount = function () {
        var reactFrameworkFactory = new reactFrameworkFactory_1.ReactFrameworkFactory(this);
        var gridParams = { frameworkFactory: reactFrameworkFactory };
        this.gridOptions = AgGrid.ComponentUtil.copyAttributesToGridOptions(this.props.gridOptions, this.props);
        AgGrid.Grid.setFrameworkBeans([reactFrameworkComponentWrapper_1.ReactFrameworkComponentWrapper]);
        // don't need the return value
        new AgGrid.Grid(this.eGridDiv, this.gridOptions, gridParams);
        this.api = this.gridOptions.api;
        this.columnApi = this.gridOptions.columnApi;
    };
    AgGridReact.prototype.shouldComponentUpdate = function () {
        // we want full control of the dom, as ag-Grid doesn't use React internally,
        // so for performance reasons we tell React we don't need render called after
        // property changes.
        return false;
    };
    AgGridReact.prototype.componentWillReceiveProps = function (nextProps) {
        var _this = this;
        var debugLogging = !!nextProps.debug;
        // keeping consistent with web components, put changing
        // values in currentValue and previousValue pairs and
        // not include items that have not changed.
        var changes = {};
        AgGrid.ComponentUtil.ALL_PROPERTIES.forEach(function (propKey) {
            if (!_this.areEquivalent(_this.props[propKey], nextProps[propKey])) {
                if (debugLogging) {
                    console.log("agGridReact: [" + propKey + "] property changed");
                }
                changes[propKey] = {
                    previousValue: _this.props[propKey],
                    currentValue: nextProps[propKey]
                };
            }
        });
        AgGrid.ComponentUtil.getEventCallbacks().forEach(function (funcName) {
            if (_this.props[funcName] !== nextProps[funcName]) {
                if (debugLogging) {
                    console.log("agGridReact: [" + funcName + "] event callback changed");
                }
                changes[funcName] = {
                    previousValue: _this.props[funcName],
                    currentValue: nextProps[funcName]
                };
            }
        });
        AgGrid.ComponentUtil.processOnChange(changes, this.gridOptions, this.api, this.columnApi);
    };
    AgGridReact.prototype.componentWillUnmount = function () {
        this.api.destroy();
    };
    /*
     * deeper object comparison - taken from https://stackoverflow.com/questions/1068834/object-comparison-in-javascript
     */
    AgGridReact.unwrapStringOrNumber = function (obj) {
        return (obj instanceof Number || obj instanceof String
            ? obj.valueOf()
            : obj);
    };
    // sigh, here for ie compatibility
    AgGridReact.prototype.copyObject = function (obj) {
        return [{}, obj].reduce(function (r, o) {
            Object.keys(o).forEach(function (k) {
                r[k] = o[k];
            });
            return r;
        }, {});
    };
    AgGridReact.prototype.areEquivalent = function (a, b) {
        return AgGridReact.areEquivalent(this.copyObject(a), this.copyObject(b));
    };
    AgGridReact.areEquivalent = function (a, b) {
        a = AgGridReact.unwrapStringOrNumber(a);
        b = AgGridReact.unwrapStringOrNumber(b);
        if (a === b)
            return true; //e.g. a and b both null
        if (a === null || b === null || typeof (a) !== typeof (b))
            return false;
        if (a instanceof Date)
            return b instanceof Date && a.valueOf() === b.valueOf();
        if (typeof (a) !== "object")
            return a == b; //for boolean, number, string, xml
        var newA = (a.areEquivalent_Eq_91_2_34 === undefined), newB = (b.areEquivalent_Eq_91_2_34 === undefined);
        try {
            var prop = void 0;
            if (newA)
                a.areEquivalent_Eq_91_2_34 = [];
            else if (a.areEquivalent_Eq_91_2_34.some(function (other) {
                return other === b;
            }))
                return true;
            if (newB)
                b.areEquivalent_Eq_91_2_34 = [];
            else if (b.areEquivalent_Eq_91_2_34.some(function (other) {
                return other === a;
            }))
                return true;
            a.areEquivalent_Eq_91_2_34.push(b);
            b.areEquivalent_Eq_91_2_34.push(a);
            var tmp = {};
            for (prop in a)
                if (prop !=
                    "areEquivalent_Eq_91_2_34")
                    tmp[prop] = null;
            for (prop in b)
                if (prop != "areEquivalent_Eq_91_2_34")
                    tmp[prop] = null;
            for (prop in tmp)
                if (!this.areEquivalent(a[prop], b[prop]))
                    return false;
            return true;
        }
        finally {
            if (newA)
                delete a.areEquivalent_Eq_91_2_34;
            if (newB)
                delete b.areEquivalent_Eq_91_2_34;
        }
    };
    return AgGridReact;
}(react_1.Component));
exports.AgGridReact = AgGridReact;
AgGridReact.propTypes = {
    gridOptions: PropTypes.object,
};
addProperties(AgGrid.ComponentUtil.getEventCallbacks(), PropTypes.func);
addProperties(AgGrid.ComponentUtil.BOOLEAN_PROPERTIES, PropTypes.bool);
addProperties(AgGrid.ComponentUtil.STRING_PROPERTIES, PropTypes.string);
addProperties(AgGrid.ComponentUtil.OBJECT_PROPERTIES, PropTypes.object);
addProperties(AgGrid.ComponentUtil.ARRAY_PROPERTIES, PropTypes.array);
addProperties(AgGrid.ComponentUtil.NUMBER_PROPERTIES, PropTypes.number);
addProperties(AgGrid.ComponentUtil.FUNCTION_PROPERTIES, PropTypes.func);
function addProperties(listOfProps, propType) {
    listOfProps.forEach(function (propKey) {
        AgGridReact[propKey] = propType;
    });
}
`

**Comment by theofanis** [comment]
I updated this, compiled my bundle and tested. The error is still there. I'll see if I can hack something together based on my actual code and reproduce it, if so, I'll share.

**Comment by seanlandsman** [comment]
thanks @theofanis

**Comment by jintoppy** [comment]
I upgraded ag-grid from version 8 to 13 and now I am getting this error

**Comment by jintoppy** [comment]
@seanlandsman I tried your changes. Now I am getting a different error. 

<img width="703" alt="screen shot 2017-08-29 at 5 47 42 pm" src="https://user-images.githubusercontent.com/3390198/29820483-37174ae6-8ce2-11e7-8063-52b0026002a8.png">

I will try to reproduce the same error in a separate repo and share soon.

**Comment by seanlandsman** [comment]
@jintoppy if you can provide a repo I'd be happy to take a look

**Comment by jintoppy** [comment]
One update on this. I was giving immutable list as column definition. Since internally, it is doing Object.freeze, it is not allowing to add any new properties. This is what was happening I believe. I converted the column definition to pure JS object before passing to AgGridReact and now it works.

**Comment by SmoooZ** [comment]
as a workaround I could get rid of this error by never rendering my AgGridReact component again.
Before v13 we used to let the component refresh in some cases (shouldComponentUpdate = true for some limited cases like when the column definition has been modified from an external module) but that goes against the Ag-grid concept of dealing only with the APIs (which goes against the React concept but this is a different story), therefore we modified the code so that the component is never rendered again (shouldComponentUpdate = false always, therefore the redux data does not need to go through this AgGridReact.prototype.componentWillReceiveProps check) and instead we detect what property has changed (via componentWillReceiveProps) and make use of the ag-grid APIs accordingly (in this specific example api.setColumnDefs()).
Hope this helps.

**Comment by jintoppy** [comment]
@SmoooZ I would suggest not to do that because, then you are not getting any benefit of using AgGridReact. You could just use AgGrid directly. 

If immutable data is the only issue(thats how my issue got fixed), when you are passing props, you just have to convert the props to normal objects rather than immutable objects. 

But, I am not sure, if immutability is the real reason for this issue, though.

**Comment by seanlandsman** [comment]
@jintoppy - thats useful, thanks - I'll see what I can do with this information

**Comment by seanlandsman** [comment]
@jintoppy - thats useful, thanks - I'll see what I can do with this information

**Comment by Kliton** [comment]
Any fix available?

**Comment by seanlandsman** [comment]
I've raised AG-771 our side to track this - you can view its progress on our pipeline page here: https://www.ag-grid.com/ag-grid-pipeline/

We'll update this thread with updates as soon as we have any.

thanks

**Comment by seanlandsman** [comment]
@jintoppy @theofanis @colvint @Kliton 

I've been playing around with this today and have tried a variety of things to try reproduce in order to fix, but have not been able to.

I'm closing this as I'm unable to reproduce this - if anyone is able to provide an example of this for me to use as a baseline I'll be happy to reopen and fix this

thanks

**Comment by amkoehler** [comment]
Same here. We're using GraphQL+Relay and the object we are passing to AgGridReact seems to be frozen. I don't know if this is something ag-grid should be responsible for handling though.

**Comment by haithemT** [comment]
same issue here using  ag-grid v 13.3.1 and react Apollo-Client  v 1.4.2

**Comment by mikeroess** [comment]
Question.  While this issue has been closed -- is there any plan to address this issue?

We use redux-freeze middleware to ensure devs new to redux do not modify the store directly (and in general to enforce good practices).

Is there any plan for ag-grid to support this, what I believe to be (could be wrong?) widespread, use case in the react-redux world?

**Comment by seanlandsman** [comment]
@mikeroess I'm happy to take a look at this again but the last time I tried i wasnt able to reproduce.

I may be missing some obvious code path - I'll try take a look at this again soon. Maybe I'll be able to reproduce it then

**Comment by mikeroess** [comment]
sorry -- am new to the tool -- it may be that we are not using enableImmutableMode, which given the name of the prop is probably required.

**Comment by deanmckee** [comment]
I'm also running into this issue using Apollo/GraphQL client with ag-grid react.  I find that I have to use Lodash cloneDeep function on the table row data (loaded from a graphql call and provided as a property to ag-grid in the render method) before I provide it to ag-grid:

      const allTopologiesCopy = lodash.cloneDeep(allTopologies);
      return (
                  <AgGridReact
                     columnDefs={this.state.columnDefs}
                     rowData={allTopologiesCopy}
                     ...
            );

**Comment by buhichan** [comment]
1 line quick fix:
```
AgGridReact.prototype.areEquivalent = (a,b) => a === b
```

**Comment by Chikel** [comment]
Facing the same issue here. Using Relay Modern (paginationContainer) with Ag-grid.
@buhichan, your solution seems to work but I fear side-effects. Can you explain how you came to resolve this?

**Comment by buhichan** [comment]
@Chikel 
AgGridReact.prototype.areEquivalent is only used in AgGridReact#componentWillReceiveProps to compare props, so I think a shallow compare is enough here:
```
AgGrid.ComponentUtil.ALL_PROPERTIES.forEach((propKey: string) => {
            if (!this.areEquivalent(this.props[propKey], nextProps[propKey])) { 
                // ⤴️ this.props[propKey] === nextProps[propKey]
                if (debugLogging) {
                    console.log(`agGridReact: [${propKey}] property changed`);
                }
                changes[propKey] = {
                    previousValue: this.props[propKey],
                    currentValue: nextProps[propKey]
                };
            }
        });
```
I can't guarantee there are no side-effects, but I haven't seen any since I used my work-around, just have to make sure not to silently mutate any deep properties of grid options.

**Comment by HydraOrc** [comment]
I have experienced to debug this issue today and my reason had not been discussed here yet.

AgGridReact took rowData, which had a property called 'style' and it was passed to the cellRendererFramework, which used it as a style property on a div:

`rowData: [
    {
        name: {
            style: {
                backgroundColor: 'red',
            }
        },
    }
]`

`<div style={style} />`

In react-dom:

`if (propKey === STYLE) {
  {
    if (nextProp) {
      // Freeze the next style object so that we can assume it won't be
      // mutated. We have already warned for this in the past.
      Object.freeze(nextProp);
    }
  }
  // Relies on `updateStylesByID` not mutating `styleUpdates`.
  setValueForStyles(domElement, nextProp, getStack);
}`

So a style property is being used on a div so it freezes for future extensions.

To solve this problem i had to change the property name to 'cellStyle' AND pass it to the div via clone:

`<div style={clone(cellStyle)} />`

**Comment by davidmoshal** [comment]
Having a similar issue with AgGridVue!
AgGridVue complains about some properties in the config:

```
gridOptionsWrapper.js:129 ag-grid: invalid colDef property '__ob__' did you mean any of these: rowGroupIndex,children,allowedAggFuncs,menuTabs,pivotTotalColumnIds,cellClassRules,icons,headerGroupComponent
```

and for rowData, complains thus:
`vue.esm.js:591 [Vue warn]: Error in render: "TypeError: Cannot read property 'state' of undefined"`

have tried lodash.cloneDeep without much luck.
In my case this was working fine before I added the Vuex-class decorators.

**Comment by biels** [comment]
This is still an issue. The AgGridReact.prototype.areEquivalent = (a,b) => a === b workaround works in my case but that doesn't mean that the issue is actually solved.

**Comment by perfect-ignorance** [comment]
Same issue using FireStore since the FireStore object is not extensible. As a dev who uses redux, isn't it my responsibility to manage component re-rendering based on object changes? Why does ag-grid have to insert itself into the middle of a functional redux update. If a deep property in one of my row-data objects changes, I am responsible for changing the reference to the parent object for the row. Three years of using ag-grid and this same area of the ag-grid code is still forcing me to do non-performant hacks like deep clones or AgGridReact.prototype overrides. :(

**Comment by filippoitaliano** [comment]
I experienced the same issue in the past and after months of doubts I realized that I was misusing the API. The `areEquivalent` function seems to NOT expect framework objects to be processed in it. So I read again the docs and I understood that the way to go is to feed the grid rowData with the "raw unformatted data" and then apply the framework renderer through column definitions or similar. Looking at the data flow in this way seems the correct way intended by ag-grid. This way is also faster (because the grid compare lighter objects), safer (because it works ignoring frameworks), and it's much simplier to support the grid export without particular hacks applying the valueGetter to the same raw data!
